### ISSUE TITLE
Support IPv6 in Sandcats.io DNS

### ISSUE THREAD

**Issue by kpreid** [sandstorm-io/sandstorm#1394]
[Asheesh Laroia wrote](https://groups.google.com/d/topic/sandstorm-dev/VInt9mdIIE4/discussion) “Sandcats does not support IPv6 at the moment”.

While we're nowhere near the point of IPv6-only systems in the world, I think that anything relating to personal servers _really ought to_ support IPv6 just in principle. And [I want to use it immediately](https://groups.google.com/d/topic/sandstorm-dev/VInt9mdIIE4/discussion).

Additionally: I couldn't tell from the docs whether or not this works already, but please make sure that the sandcats.io dynamic DNS updater prefers the `BIND_IP` of the Sandstorm server rather than any other interface the host might happen to have.

**Comment by zarvox** [comment]
To deal with NAT, sandcats.io uses the address that the Sandstorm server appears as to the sandcats.io server.  I guess the thing that would make the most sense to do would be to prefer the BIND_IP as the source address for any outgoing sockets on the Sandstorm side.

**Comment by kpreid** [comment]
Yes, if that's how it works, `bind()`ing to `BIND_IP` for the sockets opened _to the Sandcats service_ should do the trick. (If you meant binding _every_ outgoing socket, that sounds like it might unnecessarily break some cases.)

**Comment by kentonv** [comment]
Note that Google Compute Engine -- where sandcats.io is hosted -- does not support ipv6, therefore a Sandcats-using Sandstorm server that uses ipv6 would still need to communicate with the Sandcats central server using ipv4, and we have no way to automatically determine whether the Sandstorm server's ip6 address is publicly routable. While we could add manual configuration options, it would mean a very different flow from what we have currently, and it would be sort of sad.

Kevin, perhaps you should bug the GCE people to support ipv6. :)

**Comment by paulproteus** [comment]
I wonder if we can set up an IPv6 tunnel while waiting for GCE to enable
IPv6. I'll do a little research on this now. One thing I have always loved
about IPv6 is the fact that tunnelbrokers historically have been a real
option, so you can meaningfully Bring Your Own Address and expect things to
work. But I know most tunnelbrokers are shutting down lately in favor of
"real" IPv6.

But yes, definitely, Google Compute Engine ought to support IPv6!

**Comment by kpreid** [comment]
> we have no way to automatically determine whether the Sandstorm server's ip6 address is publicly routable.

If someone writes `BIND_IP=<an IPv6 address>`, isn't that a pretty good indication that that's what the DNS should resolve to? Unless they're using NAT on IPv6, I suppose, but ew.

So, an interim solution is: If bound to IPv6, send the specified address, otherwise use the observed-on-the-network address. A less magical but more heavyweight interim solution would be to add a config option.

On Google Compute Engine and IPv6, I of course can't _say_ anything more than:

> [Compute Engine currently does not support IPv6. However, Google is a major advocate of IPv6 and it is an important future direction.](https://cloud.google.com/compute/docs/networks-and-firewalls#networks)

**Comment by paulproteus** [comment]
I agree with what you said, though at the moment, the sandcats protocol has
no way for a client to express its desired IP address.

Off-topic: I did literally laugh out loud at your GCE IPv6 copy-pasta!

**Comment by kpreid** [comment]
Ping: Now that #1393 is fixed, fixing this is actually useful. As I mentioned above, better-than-nothing implementations are possible despite GCE.

**Comment by paulproteus** [comment]
Good to know!​

**Comment by kpreid** [comment]
Ping.

**Comment by kentonv** [comment]
Hi Kevin,

Since Sandstorm no longer has full-time developers, we'll need your help to implement this.

The sandcats server code is here: https://github.com/sandstorm-io/sandcats
